atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/AbstractRequestExceptionHandler.java

```diff
@@ -0,0 +1,203 @@
+        } else {
+            log.info("Error occurred talking to cassandra. Attempt {} of {}.",
+                    SafeArg.of("numTries", numberOfAttempts),
+                    SafeArg.of("maxTotalTries", maxTriesTotal),
+                    ex);
+        }
+    }
+
+    boolean shouldBlacklist(Exception ex, int numberOfAttempts) {
+        return isConnectionException(ex) && numberOfAttempts >= maxTriesSameHost.get();
+    }
+
+    <K extends Exception> void handleBackoff(RetryableCassandraRequest<?, K> req,
+            InetSocketAddress hostTried,
+            Exception ex) {
+        if (!shouldBackoff(ex)) {
+            return;
+        }
+
+        long sleepDuration = getBackoffPeriod(req.getNumberOfAttempts());
+        log.info("Retrying a query, {}, with backoff of {}ms, intended for host {}.",
+                UnsafeArg.of("queryString", req.getFunction().toString()),
+                SafeArg.of("sleepDuration", sleepDuration),
+                SafeArg.of("hostName", CassandraLogHelper.host(hostTried)));
+
+        try {
+            Thread.sleep(sleepDuration);
+        } catch (InterruptedException i) {
+            throw new RuntimeException(i);
+        }
+    }
+
+    abstract boolean shouldBackoff(Exception ex);
+
+    abstract long getBackoffPeriod(int numberOfAttempts);
+
+    <K extends Exception> void handleRetryOnDifferentHosts(RetryableCassandraRequest<?, K> req,
+            InetSocketAddress hostTried, Exception ex) {
+        if (shouldRetryOnDifferentHost(ex, req.getNumberOfAttempts())) {
+            log.info("Retrying with on a different host a query intended for host {}.",
+                    SafeArg.of("hostName", CassandraLogHelper.host(hostTried)));
+            req.giveUpOnPreferredHost();
+        }
+    }
+
+    abstract boolean shouldRetryOnDifferentHost(Exception ex, int numberOfAttempts);
+
+    // Determine the behavior we want from each type of exception.
+
+    @VisibleForTesting
+    boolean isRetryable(Exception ex) {
+        return isConnectionException(ex)
+                || isTransientException(ex)
+                || isIndicativeOfCassandraLoad(ex)
+                || isFastFailoverException(ex);
+    }
+
+    // Group exceptions by type.
+
+    static boolean isConnectionException(Throwable ex) {
+        return ex != null
+                // tcp socket timeout, possibly indicating network flake, long GC, or restarting server.
+                && (ex instanceof SocketTimeoutException
+                || ex instanceof CassandraClientFactory.ClientCreationFailedException
+                || isConnectionException(ex.getCause()));
+    }
+
+    private static boolean isTransientException(Throwable ex) {
+        return ex != null
+                // There's a problem with the connection to Cassandra.
+                && (ex instanceof TTransportException
+                // Not enough Cassandra nodes are up.
+                || ex instanceof InsufficientConsistencyException
```

Comment:
I think we want to have the same behavior for `InsufficientConsistencyException` and `UnavailableException`, since we try to wrap the latter in the former in several places.

Reply:
Agree, also we want to backoff when we don't have enough nodes up